### ISSUE TITLE
How to distinguish user-initiated map move events from events fired as a result of programmatic APIs?

### ISSUE THREAD

**Issue by igalgh** [maplibre/maplibre-gl-js#793]
**maplibre-gl-js version**: (all)

### Question
I need a way to respond differently to map move/pan/pitch/bearing events depending on those events being a result of user-interaction or programmatic API calls. 

Reasoning: 
We have a special mode where map view changes which are a result of user interactions are written to the DB. 
Map movements which is a result of API calls do not.

One way to implement it is to add an `eventData` argument to all programmatic API calls as in:
```js 
map.setCenter(center, {originIsApi: true}) 
```
It can hold a flag which implies the event originates from an API call and not a result of user interaction. 

This approach seems a little hackish thou. It requires updating every api call in the project: force an `eventData` arg for each, which is only used for this reason.

It'd be so much more elegant to have this information built into the events by design. 
It'll save the developer the effort of modifying the entire API calls set.  

The `eventData` approach also involves an elevated risk for bugs; inexperienced developers could neglect providing a suitable `eventData`arg, introducing very subtle bugs which are difficult to track down.

Anyone else facing a similar issue? Perhaps I'm missing a better solution...

**Comment by igal1c0de4n** [comment]
This problem is, the user has many ways to modify the maps. With the mouse, with touch, with keyboard (affecting pitch, bearing etc). It seems a little overwhelming to try and catch all of those different methods. 
Also there's always the lingering "what if I missed an event?" concern, which even gets one to be weary about using new versions since those may have added events which aren't covered by the current app code

**Comment by HarelM** [comment]
Yea, I can't argue here, but I think implementing is as brittle as using it for the same reasons :-)

**Comment by igal1c0de4n** [comment]
IDK. 
Zooming out on the subject, asking "Is the user responsible for this event?"  
seems an important enough question to deserve a native answer

**Comment by igal1c0de4n** [comment]
I wonder if I can rely on the existence of `originalEvent`, the first argument to the `moveend` callback.
It's missing if the the map view changes via API call, but it seems to be almost always there on events originating at user interaction except for the events originating at the bearing control widget

**Comment by zhangyiatmicrosoft** [comment]
> I wonder if I can rely on the existence of `event.originalEvent`, where `event` is the first argument to the `moveend` callback.
> 
> `originalEvent` is missing if the the map view changes via API call however it seems to exist on user-interaction events, except for bearing-control widget events

Yes that is a pretty stable solution and works for mouse (drag and drop), keyboard, touch --- EXCEPT mouse wheel which triggers zoom.

originalEvent field is missing in wheel

**Comment by johncarmack1984** [comment]
originalEvent is undefined on programmatic moves and set on user ones; the thread settled on that in 2024. Think this one is closable.
